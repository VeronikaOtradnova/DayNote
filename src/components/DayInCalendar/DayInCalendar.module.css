.cell {
  padding: 10px 0;
  color: #3B3D49;
}
.cell.weekend {
  color: #AE1314;
}

.contentWrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 4rem;
  height: 4rem;
  margin: auto;
  padding: 5px;
  border-radius: 5px;
  border: 3px solid transparent;
}

.selected.green {
  border-color: var(--green);
}
.selected.yellow {
  border-color: var(--yellow);
}
.selected.gray {
  border-color: var(--gray);
}

.number {
  border-bottom: 3px solid var(--gray);
}
.number.green {
  border-color: var(--green);
}
.number.yellow {
  border-color: var(--yellow);
}

@media (min-width: 500px) {
  .cell {
    padding: 10px 5px;
  }

  .contentWrapper {
    width: 5rem;
    height: 5rem;
  }
}

@media (min-width: 1000px) {
  .cell {
    padding: 10px 10px;
  }
}