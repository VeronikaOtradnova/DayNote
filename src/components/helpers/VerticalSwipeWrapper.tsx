import { useState } from "react"

interface IProps {
  children: React.ReactNode,
  onTop?: () => void,
  onBottom?: () => void,
}

export function VerticalSwipeWrapper({children, onTop = () => {}, onBottom = () => {}}: IProps) {
  const [touchStart, setTouchStart] = useState<number | null>(null)
  const [touchEnd, setTouchEnd] = useState<number | null>(null)

  // the required distance between touchStart and touchEnd to be detected as a swipe
  const minSwipeDistance = 50

  const onTouchStart = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(null)
    setTouchStart(e.targetTouches[0].clientY)
  }

  const onTouchMove = (e: React.TouchEvent<HTMLDivElement>) => {
    setTouchEnd(e.targetTouches[0].clientY);
  };

  const onTouchEnd = () => {
    if (!touchStart || !touchEnd) return
    const distance = touchStart - touchEnd
    const isTopSwipe = distance > minSwipeDistance
    const isBottomSwipe = distance < -minSwipeDistance
    if (isTopSwipe || isBottomSwipe) {
      isTopSwipe ? onTop() : onBottom()
    }
  }

  return (
    <div 
      onTouchStart={onTouchStart} 
      onTouchMove={onTouchMove} 
      onTouchEnd={onTouchEnd} 
    >
      {children}
    </div>
  )
}