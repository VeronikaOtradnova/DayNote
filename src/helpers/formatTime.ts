export function formatTime(timeInMillis: number) {
  const date = new Date(timeInMillis);
  
  // Получаем часы и минуты из объекта Date
  let hours = date.getUTCHours().toString();
  let minutes = date.getUTCMinutes().toString();
  
  // Форматируем часы и минуты в виде двухзначных чисел
  hours = hours.padStart(2, '0');
  minutes = minutes.padStart(2, '0');
  
  // Возвращаем строку в формате 'hh:mm'
  return `${hours}:${minutes}`;
}