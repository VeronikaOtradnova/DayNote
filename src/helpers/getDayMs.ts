export const getDayMs = () => {
  return 24*60*60*1000;
}