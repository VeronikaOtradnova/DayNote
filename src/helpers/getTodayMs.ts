export const getTodayMs = () => {
  return +(new Date().setHours(0, 0, 0, 0));
}