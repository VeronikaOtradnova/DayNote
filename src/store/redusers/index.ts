import { combineReducers } from "redux";
import { dayReducer } from "./dayReducer";
import { calendarReducer } from "./calendarReducer";

export const rootReducer = combineReducers({
  day: dayReducer,
  calendar: calendarReducer
})

export type TRootState = ReturnType<typeof rootReducer>