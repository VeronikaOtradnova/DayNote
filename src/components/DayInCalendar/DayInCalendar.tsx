import React from 'react';
import { generateString } from '../../helpers/generateString';
import { useActions } from '../../hooks/useActions';
import styles from './DayInCalendar.module.css';
import { useTypedSelector } from '../../hooks/useTypedSelector';
import { colors } from '../../types/color';
import { getTodayMs } from '../../helpers/getTodayMs';

interface IProps {
  date: Date
};

export const DayInCalendar = ({ date }: IProps) => {
  const { calendarDate } = useTypedSelector(store => store.calendar);
  const { days } = useTypedSelector(store => store.day);
  const { setCalendarDate } = useActions();

  const isDayInFuture = +date > getTodayMs() ? true : false;
  const isWeekend = (date.getDay() === 0 || date.getDay() === 6) ? true : false;
  const cellClassName = isWeekend ? `${styles.cell} ${styles.weekend}` : styles.cell;

  const isSelectedDate = +date === calendarDate;
  let color = days.find(d => d.date === +date)?.color || colors.gray;

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    if (!isSelectedDate) {
      setCalendarDate(+date);
    }
  }

  return (
    <>
      {
        isDayInFuture ?
          <td className={cellClassName} key={generateString()} data-testid='day-in-calendar_future'>
            <div className={styles.contentWrapper}>
              {date.getDate()}
            </div>
          </td>
          :
          <td onClick={handleClick} className={cellClassName} key={generateString()} data-testid='day-in-calendar'>
            <div className={isSelectedDate ? `${styles.contentWrapper} ${styles.selected} ${styles[color]}` : styles.contentWrapper}>
              <span className={`${styles.number} ${styles[color]}`}>{date.getDate()}</span>
            </div>
          </td>
      }
    </>
  )
}