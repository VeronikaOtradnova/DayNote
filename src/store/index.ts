import { applyMiddleware, createStore } from "redux";
import { rootReducer } from "./redusers";
import { thunk } from "redux-thunk";

const stateFromStorage = localStorage.getItem('reduxState')
export const persistedState = stateFromStorage ? JSON.parse(stateFromStorage): {}

export const store = createStore(rootReducer, persistedState, applyMiddleware(thunk));

export type AppDispatch = typeof store.dispatch;

store.subscribe(()=>{
  localStorage.setItem('reduxState', JSON.stringify(store.getState()))
})