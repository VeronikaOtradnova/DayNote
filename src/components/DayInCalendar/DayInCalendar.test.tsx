import { screen } from "@testing-library/react";
import { TRootState } from "../../store/redusers";
import { renderWithRedux, testInitialState } from "../../tests/helpers/renderWithRedux";
import { DayInCalendar } from "./DayInCalendar";
import userEvent from "@testing-library/user-event";
import { getTodayMs } from "../../helpers/getTodayMs";
import { getDayMs } from "../../helpers/getDayMs";
import { act } from "react-dom/test-utils";

const initialState: TRootState = {
  ...testInitialState,
  day: {
    currentDay: +(new Date(2020, 0, 10)),
    days: []
  },
  calendar: {
    isCalendarOpen: true,
    calendarDate: +(new Date(2020, 0, 10)),
  },
}

describe('TEST DAY IN CALENDAR', () => {
  test('Should render not selected day', () => {
    renderWithRedux(<DayInCalendar date={new Date(2020, 0, 1)} />, { initialState: initialState });

    const dayElem = screen.getByTestId('day-in-calendar');
    expect(dayElem).toBeInTheDocument();
    expect(dayElem).toHaveTextContent('1');
    expect(dayElem).toMatchSnapshot();
  })

  test('Should set calendarDate after click', async () => {
    const user = userEvent.setup();
    const {store} = renderWithRedux(<DayInCalendar date={new Date(2020, 0, 1)} />, { initialState: testInitialState });

    const dayElem = screen.getByTestId('day-in-calendar');
    await user.click(dayElem);
    expect(dayElem).toMatchSnapshot();

    const state:TRootState = store.getState();
    expect(state.calendar.calendarDate).toBe(+new Date(2020, 0, 1));
  })

  test('Should not set calendarDate after click on future-day', async () => {
    const user = userEvent.setup();
    const {store} = renderWithRedux(<DayInCalendar date={new Date(getTodayMs() + getDayMs())} />, { initialState: testInitialState });
    const stateBeforeClick:TRootState = store.getState();

    expect(screen.queryByTestId('day-in-calendar')).not.toBeInTheDocument();
    const dayElem = screen.getByTestId('day-in-calendar_future');
    expect(dayElem).toBeInTheDocument();

    await act(async () => {
      await user.click(dayElem);
    })

    const state:TRootState = store.getState();
    expect(state.calendar.calendarDate).toBe(stateBeforeClick.calendar.calendarDate);
  })
})