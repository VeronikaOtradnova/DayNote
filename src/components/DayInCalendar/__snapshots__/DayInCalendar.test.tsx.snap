// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST DAY IN CALENDAR Should not set calendarDate after click on future-day 1`] = `
<td
  class="cell"
  data-testid="day-in-calendar_future"
>
  <div
    class="contentWrapper"
  >
    23
  </div>
</td>
`;

exports[`TEST DAY IN CALENDAR Should render not selected day 1`] = `
<td
  class="cell"
  data-testid="day-in-calendar"
>
  <div
    class="contentWrapper"
  >
    <span
      class="number gray"
    >
      1
    </span>
  </div>
</td>
`;

exports[`TEST DAY IN CALENDAR Should set calendarDate after click 1`] = `
<td
  class="cell"
  data-testid="day-in-calendar"
>
  <div
    class="contentWrapper"
  >
    <span
      class="number gray"
    >
      1
    </span>
  </div>
</td>
`;
